lib/dynflow/director.rb

```diff
@@ -146,14 +146,20 @@ def finish_manager(manager)
 
     def rescue?(manager)
       return false if @world.terminating?
-      @world.auto_rescue && manager.execution_plan.state == :paused &&
-        !@plan_ids_in_rescue.include?(manager.execution_plan.id)
+      return false unless @world.auto_rescue && manager.execution_plan.state == :paused
+      # we have not rescued this plan yet
+      return true unless @rescued_steps.key?(manager.execution_plan.id)
+      # we have rescued this plan already, but a different step has failed now
+      # we do this check to prevent endless loop, if we always failed on the same steps
+      failed_step_ids = manager.execution_plan.failed_steps.map(&:id).to_set
+      return true unless (failed_step_ids - @rescued_steps[manager.execution_plan.id]).empty?
```

Comment:
Wouldn't this be more readable rewritten as a if-elsif-else? I'm not really all that fond of multiple early returns.
```ruby
if @world.terminating? || !(@world.auto_rescue && manager.execution_plan.state == :paused)
  false
elsif !@rescued_steps.key?(manager.execution_plan.id)
  # we have not rescued this plan yet
  true
else
  # we have rescued this plan already, but a different step has failed now
  # we do this check to prevent endless loop, if we always failed on the same steps
  failed_step_ids = manager.execution_plan.failed_steps.map(&:id).to_set
  !(failed_step_ids - @rescued_steps[manager.execution_plan.id]).empty?
end
```